### ISSUE TITLE
Structure Matcher

### ISSUE THREAD

**Issue by legendblade** [legendblade/CraftingHarmonics#76]
Add matchers to check if events occur at or around a specified structure.

Something like:
```
nearStructure: {
   structure: [
      [[' ','B',' ']],
      [['A','A','A']]
   ],
   blocks: [
      A: ["minecraft:diamond_block"],
      B: ["minecraft:soul_sand"]
   ],
   dist: 5,
   validRotations: ['N','S','E',W']
}
```

Or:
```
inStructure: {
   structure: [
      [
         ['B','A','B']
         ['B','C','B']
         ['B','B','B']
      ],
      [
         ['A','A','A']
         ['A','D','A']
         ['A','A','A']
      ],
   ],
   blocks: [
      A: ["minecraft:diamond_block"],
      B: ["minecraft:soul_sand","minecraft:dirt"],
      C: ["minecraft:air],
      D: [{
         block: "minecraft:dirt",
         state: {
            variant: "podzol"
         }
      }]
   ],
   validPositions: ['C']
}
```

- The first would detect if you were within 5 blocks of a series of three diamond blocks, with a single piece of soul sand placed on top of them, along any orientation.
- The second would detect if the triggering entity was specifically in the center of:
   - Bottom layer:
      - A ring of diamond blocks
      - A single piece of podzol in the middle
   - Top layer:
      - A ring of soul sand or dirt
      - With a single diamond block on the north face
      - Air in the center (this would restrict this to being a mob drop / mob shed / entity effect, or, once implemented, a block placement, since, if you were doing this for a block break, you'd want 'C' to be the type of block you wanted to be broken)

**Comment by legendblade** [comment]
@Thiana since you mentioned detecting players being near shrines in a particular issue, would something like this be of use to you?

**Comment by Thiana** [comment]
Currently the shrine 'detection' is based on the player being in proximity to a shrine (location trigger). But I can see this idea being *very* useful in general. It would allow the pack author to create tiers of trigger structures. ie, a small structure gives a small effect (blessing/shed bonus/etc) but as they create larger, more complex, structures the effect could increase.

My concern about this type of matcher though would be the efficiency. Large numbers of block scans probably aren't that efficient in terms of CPU?

**Comment by legendblade** [comment]
Yeah. I was considering that as well.  My thought process went much like so:
- "Hey, this could be a thing" 
- "Alright, I'll define it and then check it when the event happens."
- "No, that's a really terrible idea." :laughing: 

Which then led me to the implementation I'll probably start out with initially, which would be to do it much like a multiblock:
- Keep track of blocks (rather: block states) that are used in any structure
- When a block is placed (also, broken, moved, etc), check if it's part of the blocks that need to be tested for being a structure (relatively quick check)
- If it is, do the actual structure check (the longer part; could see about passing this off to a separate thread for initial validation)
- If a valid structure is found, save that in a list somewhere.

Then, when an event occurs with a structure matcher, check to see if it happened near one of those structures, and, if so, do the actual event check.

I *think* that would save enough CPU cycles, provided the pack creator doesn't go completely overboard.  But then it's their problem. :smirk:
